crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move

```diff
@@ -0,0 +1,252 @@
+        assert!(is_installed<Ext>(self), EExtensionNotInstalled);
+        extension_mut<Ext>(self).is_enabled = false;
+    }
+
+    /// Re-enable the extension allowing it to call protected actions (eg
+    /// `place`, `lock`). By default, all added extensions are enabled. Kiosk
+    /// owner can disable them via `disable` call.
+    public fun enable<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+    ) {
+        assert!(kiosk::has_access(self, cap), ENotOwner);
+        assert!(is_installed<Ext>(self), EExtensionNotInstalled);
+        extension_mut<Ext>(self).is_enabled = true;
+    }
+
+    /// Remove an extension from the Kiosk. Can only be performed by the owner,
+    /// the extension storage must be empty for the transaction to succeed.
+    public fun remove<Ext: drop>(
+        self: &mut Kiosk, cap: &KioskOwnerCap
+    ) {
+        assert!(kiosk::has_access(self, cap), ENotOwner);
+        assert!(is_installed<Ext>(self), EExtensionNotInstalled);
+
+        let Extension {
+            storage,
+            permissions: _,
+            is_enabled: _,
+        } = df::remove(kiosk::uid_mut_as_owner(self, cap), ExtensionKey<Ext> {});
+
+        bag::destroy_empty(storage);
+    }
+
+    // === Storage ===
+
+    /// Get immutable access to the extension storage. Can only be performed by
+    /// the extension as long as the extension is installed.
+    public fun storage<Ext: drop>(
+        _ext: Ext, self: &Kiosk
+    ): &Bag {
+        assert!(is_installed<Ext>(self), EExtensionNotInstalled);
+        &extension<Ext>(self).storage
+    }
+
+    /// Get mutable access to the extension storage. Can only be performed by
+    /// the extension as long as the extension is installed. Disabling the
+    /// extension does not prevent it from accessing the storage.
+    ///
+    /// Potentially dangerous: extension developer can keep data in a Bag
+    /// therefore never really allowing the KioskOwner to remove the extension.
+    /// However, it is the case with any other solution (1) and this way we
+    /// prevent intentional extension freeze when the owner wants to ruin a
+    /// trade (2) - eg locking extension while an auction is in progress.
+    ///
+    /// Extensions should be crafted carefully, and the KioskOwner should be
+    /// aware of the risks.
+    public fun storage_mut<Ext: drop>(
+        _ext: Ext, self: &mut Kiosk
+    ): &mut Bag {
+        assert!(is_installed<Ext>(self), EExtensionNotInstalled);
+        &mut extension_mut<Ext>(self).storage
+    }
+
+    // === Protected Actions ===
+
+    /// Protected action: place an item into the Kiosk. Can be performed by an
+    /// authorized extension. The extension must have the `place` permission
+    /// and the type of the item must be in the list of allowed types.
+    ///
+    /// To prevent non-tradable items from being placed into `Kiosk` the method
+    /// requires a `TransferPolicy` for the placed type.
```

Comment:
```suggestion
    /// requires a `TransferPolicy` for the placed type to exist.
```

What happens if the policy exists at the time of placing and is subsequently made inaccessible somehow?